### PR TITLE
Add RAML

### PR DESCRIPTION
**Pull request by eddumelendez**
- [X] Convert Raml to Contract
- [ ]  Convert Contract to Raml
- [ ] Add docs

See #129

### CODE REVIEW

File: spring-cloud-contract-tools/pom.xml [hunk 1]
```diff
@@ -21,7 +21,8 @@
 		<module>spring-cloud-contract-converters</module>
 		<module>spring-cloud-contract-spec-pact</module>
 		<module>spring-cloud-contract-maven-plugin</module>
-		<module>spring-cloud-contract-gradle-plugin</module>
+		<!--<module>spring-cloud-contract-gradle-plugin</module>-->
```
Comment (marcingrzejszczak) [inline spring-cloud-contract-tools/pom.xml]: lol, why? Too slow? :P You can run with `-Pfast` to ignore the tests from that module

File: spring-cloud-contract-verifier/pom.xml [hunk 2]
```diff
@@ -74,6 +74,10 @@
 			<artifactId>groovy-xml</artifactId>
 		</dependency>
 		<dependency>
+			<groupId>org.raml</groupId>
+			<artifactId>raml-parser-2</artifactId>
+		</dependency>
```
Comment (marcingrzejszczak) [inline spring-cloud-contract-verifier/pom.xml]: Mayb let's make this optional?

File: spring-cloud-contract-verifier/src/main/resources/META-INF/spring.factories [hunk 3]
```diff
@@ -5,4 +5,8 @@ org.springframework.cloud.contract.verifier.messaging.integration.ContractVerifi
 org.springframework.cloud.contract.verifier.messaging.amqp.ContractVerifierAmqpAutoConfiguration,\
 org.springframework.cloud.contract.verifier.messaging.amqp.RabbitMockConnectionFactoryAutoConfiguration,\
 org.springframework.cloud.contract.verifier.messaging.camel.ContractVerifierCamelConfiguration,\
-org.springframework.cloud.contract.verifier.messaging.noop.NoOpContractVerifierAutoConfiguration
\ No newline at end of file
+org.springframework.cloud.contract.verifier.messaging.noop.NoOpContractVerifierAutoConfiguration
+
+# Converters
+org.springframework.cloud.contract.spec.ContractConverter=\
+org.springframework.cloud.contract.verifier.converter.RamlContractConverter
```
Comment (marcingrzejszczak) [inline spring-cloud-contract-verifier/src/main/resources/META-INF/spring.factories]: What's this?

File: spring-cloud-contract-tools/spring-cloud-contract-spec-raml/src/main/groovy/org/springframework/cloud/contract/verifier/spec/raml/RamlContractConverter.groovy [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    @Override
+    Collection<Contract> convertFrom(File file) {
+        RamlModelBuilder ramlBuilder = new RamlModelBuilder()
+        RamlModelResult api = ramlBuilder.buildApi(file)
+        return api.apiV10.resources().collect { Resource resource ->
+            Contract.make {
+                description(resource.description().value())
+                request {
+                    url(resource.relativeUri().value()) {
+                        queryParameters {
```
Comment (marcingrzejszczak) [inline spring-cloud-contract-tools/spring-cloud-contract-spec-raml/src/main/groovy/org/springframework/cloud/contract/verifier/spec/raml/RamlContractConverter.groovy]: Maybe let's add query params only if there are some?

File: spring-cloud-contract-tools/spring-cloud-contract-spec-raml/src/main/groovy/org/springframework/cloud/contract/verifier/spec/raml/RamlContractConverter.groovy [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+                    }
+
+                    resource.methods().each { Method httpMethod ->
+                        method(httpMethod.method().toUpperCase())
+
+                        httpMethod.body().each { TypeDeclaration requestBody ->
+                            JsonSlurper parser = new JsonSlurper()
+                            if (requestBody.defaultValue()) {
+                                def parsedBody = parser.parseText(requestBody.defaultValue())
+                                if (parsedBody instanceof Map) {
```
Comment (marcingrzejszczak) [inline spring-cloud-contract-tools/spring-cloud-contract-spec-raml/src/main/groovy/org/springframework/cloud/contract/verifier/spec/raml/RamlContractConverter.groovy]: I guess this whole section can be extracted and reused

```
if (parsedBody instanceof Map) {
 +                                    body(parsedBody as Map)
 +                                } else if (parsedBody instanceof List) {
 +                                    body(parsedBody as List)
 +                                } else {
 +                                    body(parsedBody.toString())
 +                                }
```

**Comment by marcingrzejszczak** [comment]
Looks really nice!! So what's the catch with the `Collection<Contract>` -> `RAML` conversion?

**Comment by marcingrzejszczak** [comment]
Ok I think I know what the problem is :P I found this lib that we can use maybe? https://github.com/mulesoft/raml-for-jax-rs, we don't care about the jax-rs part so we can completely exclude it. We can also copy paste their model and parser and reference it

**Comment by marcingrzejszczak** [comment]
Ok I know completely know why you said that it's problematic ;) I've done some of this but I came to the body and now I'm think that we have to choices:

- We'd have to parse it and instead of a value for each node create a type node that defines what that type is
- just set an explicit type like "any"

I actually am in favour of the latter. That's because SC-Contract and contracts as such are not tools to design schemas. Actually what we would present is not really a schema. It's a subset of it. For some M1 I'd suggest we present an MVP without parsing bodies and transforming them into a RAML acceptable structure.

I've pushed my changes here - https://github.com/spring-cloud/spring-cloud-contract/tree/eddumelendez-gh-129 . If you want to continue with this just pick my changes :P

WDYT @eddumelendez ?

**Comment by marcingrzejszczak** [comment]
I have just seen that there's a field in RAML called `example`. I guess that's our only way to pass some request / reply simulation

**Comment by eddumelendez** [comment]
Thanks for your help and suggestions @marcingrzejszczak, I'll take a look and hope to have some deliverable soon.

**Comment by marcingrzejszczak** [comment]
@eddumelendez I sat down a little bit and managed to get some results. You can check out my branch (it's not refactored so it doesn't loot beautiful and has some hardcoded values cause I didn't know what should go in those places) for more info.

**Comment by marcingrzejszczak** [comment]
Closing this in favour of https://github.com/spring-cloud/spring-cloud-contract/pull/227

**Pull request closed without merging**